htmresearch/frameworks/layers/l2_l4_network_creation.py

```diff
@@ -333,3 +359,74 @@ def createMultipleL4L2Columns(network, networkConfig):
   enableProfiling(network)
 
   return network
+
+
+def createMultipleL4L2ColumnsWithTopology(network, networkConfig):
+    """
+    Create a network consisting of multiple columns.  Each column contains one
+    L4 and one L2, is identical in structure to the network created by
+    createL4L2Column. In addition the L2 columns are connected to each
+    other through their lateral inputs, based on the topological information
+    provided.
+
+    Region names have a column number appended as in externalInput_0,
+    externalInput_1, etc.
+
+    networkConfig must be of the following format (see createL4L2Column for
+    further documentation):
+
+      {
+        "networkType": "MultipleL4L2Columns",
+        "numCorticalColumns": 3,
+        "externalInputSize": 1024,
+        "sensorInputSize": 1024,
+        "columnPositions": a list of 2D coordinates, one for each column.
+        Used to calculate the connections between columns. By convention,
+        coordinates should be integers.
+        "maxConnectionDistance": should be a value >= 1.  Determines how distant
+        of columns will be connected to each other.  Useful specific values
+        are 1 and 1.5, which typically create grids without and with diagonal
+        connections, respectively.
+        "L4Params": {
+          <constructor parameters for ExtendedTMRegion>
+        },
+        "L2Params": {
+          <constructor parameters for ColumnPoolerRegion>
+        },
+        "lateralSPParams": {
+          <constructor parameters for optional SPRegion>
+        },
+        "feedForwardSPParams": {
+          <constructor parameters for optional SPRegion>
+        }
+      }
+    """
+
+    # Create each column
+    numCorticalColumns = networkConfig["numCorticalColumns"]
+    for i in xrange(numCorticalColumns):
+      networkConfigCopy = copy.deepcopy(networkConfig)
+      layerConfig = networkConfigCopy["L2Params"]
+      layerConfig["seed"] = layerConfig.get("seed", 42) + i
+
+      layerConfig["numOtherCorticalColumns"] = numCorticalColumns - 1
+
+      suffix = "_" + str(i)
+      network = createL4L2Column(network, networkConfigCopy, suffix)
+
+    # Now connect the L2 columns laterally
+    for i, src_pos in enumerate(networkConfig["columnPositions"]):
+      suffixSrc = "_" + str(i)
+      for j, dest_pos in enumerate(networkConfig["columnPositions"]):
```

Comment:
Can we avoid iterating over all column positions and instead figure out the 2D square of potential columns that could be connected? As coded, this is `n^2` with n being number of columns where it could just be `n*c` with c being the `maxConnectionDistance` squared